workflow/util/util.go

```diff
@@ -827,6 +827,29 @@ func resetConnectedParentGroupNodes(oldWF *wfv1.Workflow, newWF *wfv1.Workflow,
 	return newWF, resetParentGroupNodes
 }
 
+// RteryWorkflow retry a workflow by setting spec.retry and controller will retry it.
+func RetryWorkflow(ctx context.Context, wf *wfv1.Workflow, restartSuccessful bool, nodeFieldSelector string, parameters []string) (*wfv1.Workflow, error) {
+	switch wf.Status.Phase {
+	case wfv1.WorkflowFailed, wfv1.WorkflowError:
+	case wfv1.WorkflowSucceeded:
+		if !(restartSuccessful && len(nodeFieldSelector) > 0) {
+			return nil, errors.Errorf(errors.CodeBadRequest, "To retry a succeeded workflow, set the options restartSuccessful and nodeFieldSelector")
+		}
+	default:
+		return nil, errors.Errorf(errors.CodeBadRequest, "Cannot retry a workflow in phase %s", wf.Status.Phase)
+	}
+
+	wf.Spec.Retry = &wfv1.RetryConfig{
+		Retried:           false,
+		RestartSuccessful: restartSuccessful,
+		NodeFieldSelector: nodeFieldSelector,
+		Parameters:        parameters,
+	}
+
+	delete(wf.Labels, common.LabelKeyCompleted)
```

Comment:
the delete label looks formulation of workflow, I think we can let `formulateRetryWorkflow` handle this instead.
so that server only 
- check client request error
- and update for retry

Reply:
If delete it from here, reconceil may not be triggered. Do you think  change "reconciliationNeeded" function better than delete label here? If so, I will modify reconciliationNeeded function. 
```
func reconciliationNeeded(wf metav1.Object) bool {
	return wf.GetLabels()[common.LabelKeyCompleted] != "true" || slices.Contains(wf.GetFinalizers(), common.FinalizerArtifactGC)
}
```